crates/iroha_genesis/src/lib.rs

```diff
@@ -71,241 +76,222 @@ impl RawGenesisTransaction {
+    /// Fails if [`RawGenesisTransaction::parse`] fails.
     pub fn build_and_sign(self, genesis_key_pair: &KeyPair) -> Result<GenesisBlock> {
-        let executor = get_executor(&self.executor.0)?;
-        let parameters = self
-            .parameters
-            .map_or(Vec::new(), |parameters| parameters.parameters().collect());
-        let genesis = build_and_sign_genesis(
-            self.instructions,
-            executor,
-            self.chain,
-            genesis_key_pair,
-            self.topology,
-            parameters,
-        );
-        Ok(genesis)
+        let chain = self.chain.clone();
+        let mut transactions = vec![];
+        for instructions in self.parse()? {
+            let transaction = build_transaction(instructions, chain.clone(), genesis_key_pair);
+            transactions.push(transaction);
+        }
+        let block = SignedBlock::genesis(transactions, genesis_key_pair.private_key());
+
+        Ok(GenesisBlock(block))
     }
-}
 
-fn build_and_sign_genesis(
-    instructions: Vec<InstructionBox>,
-    executor: Executor,
-    chain_id: ChainId,
-    genesis_key_pair: &KeyPair,
-    topology: Vec<PeerId>,
-    parameters: Vec<Parameter>,
-) -> GenesisBlock {
-    let transactions = build_transactions(
-        instructions,
-        executor,
-        parameters,
-        topology,
-        chain_id,
-        genesis_key_pair,
-    );
-    let block = SignedBlock::genesis(transactions, genesis_key_pair.private_key());
-    GenesisBlock(block)
-}
+    /// Parse [`RawGenesisTransaction`] to the list of source instructions of the genesis transactions
+    ///
+    /// # Errors
+    ///
+    /// Fails if `self.executor` path fails to load [`Executor`].
+    fn parse(self) -> Result<Vec<Vec<InstructionBox>>> {
+        let mut instructions_list = vec![];
 
-fn build_transactions(
-    instructions: Vec<InstructionBox>,
-    executor: Executor,
-    parameters: Vec<Parameter>,
-    topology: Vec<PeerId>,
-    chain_id: ChainId,
-    genesis_key_pair: &KeyPair,
-) -> Vec<SignedTransaction> {
-    let upgrade_isi = Upgrade::new(executor).into();
-    let transaction_executor =
-        build_transaction(vec![upgrade_isi], chain_id.clone(), genesis_key_pair);
-    let mut transactions = vec![transaction_executor];
-    if !topology.is_empty() {
-        let register_peers = build_transaction(
-            topology
+        let upgrade_executor = Upgrade::new(self.executor.try_into()?).into();
+        instructions_list.push(vec![upgrade_executor]);
+
+        if let Some(parameters) = self.parameters {
+            let instructions = parameters
+                .parameters()
+                .map(SetParameter::new)
+                .map(InstructionBox::from)
+                .collect();
+            instructions_list.push(instructions);
+        }
+
+        if !self.instructions.is_empty() {
+            instructions_list.push(self.instructions);
+        }
+
+        if !self.topology.is_empty() {
```

Comment:
Topology is processed last. No gap between the order seen in `genesis.json` and the actual process